flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java

```diff
@@ -65,9 +64,15 @@
 	private final Map<ObjectPath, Map<CatalogPartitionSpec, CatalogColumnStatistics>> partitionColumnStats;
 
 	public GenericInMemoryCatalog(String name) {
+		this(name, DEFAULT_DB);
+	}
+
+	public GenericInMemoryCatalog(String name, String defaultDatabase) {
```

Comment:
How about `GenericInMemoryCatalog(String name, String defaultDatabaseName, CatalogDatabase defaultDatabase)`? I guess we don't need to register additional database if user wants to set a different one.

I think we don't need to check for db existence in the `open` method if we do that.

Reply:
Hi @dawidwys Thanks for your review. To clarify, do you mean we don't want user to set a  default database other than DEFAULT_DB? I'm fine with this because there is the only one it has when the catalog is initialized.

Even for GenericInMemoryCatlog, it confirms the catalog API, allowing user to create different databases.

Reply:
No, what I mean is to allow user to provide the defaultDB. In other words to have a ctor that does not create the defaultDb, but uses the one provided by the user.

So the ctor would like:
```
	public GenericInMemoryCatalog(String name, String defaultDbName, CatalogDatabase defaultDb) {
		checkArgument(!StringUtils.isNullOrWhitespaceOnly(name), "name cannot be null or empty");

		this.currentDatabase = defaultDbName;
		this.catalogName = name;
		this.databases = new LinkedHashMap<>();
		this.databases.put(defaultDbName, defaultDb);
		this.tables = new LinkedHashMap<>();
		this.functions = new LinkedHashMap<>();
		this.partitions = new LinkedHashMap<>();
		this.tableStats = new LinkedHashMap<>();
		this.tableColumnStats = new LinkedHashMap<>();
		this.partitionStats = new LinkedHashMap<>();
		this.partitionColumnStats = new LinkedHashMap<>();
	}
```

Reply:
Cool. Got it. Will update. Thanks.

Reply:
+1, I like Xuefu's approach as it conforms more to yaml config.

Reply:
Hmm, not sure if it conforms to yaml more. How do you define database properties in the yaml file? How do you set the properties of the default db in yaml?

I could imagine sth like:
```
catalogs:
  - name: ABC
    type: in-memory
    property-version: 1
    default-database: 
        name: mydb
        properties:
            prop1: "prop1"
            ....
```

Reply:
I'm not in favor of that for a couple reasons. 1) Currently yaml file is misused as some kind of "file-based catalog" where users can define metadata like tables, views, etc. The effort of unified catalog API is to correct that misuse by moving metadata out and only keep configuration values in yaml. The above proposed format in which allows users to define database would lead us back to the misuse case. 2)  in-memory catalog is mainly used for back-compatibility and when users want to experiment with Flink. We don't expect/recommend users to set an in-memory catalog in yaml file besides the default one. 

To take a step back, I think we can clearly define that users can only set default db for a persistent catalog, and not in-memory catalog, as no db except the "build-in" db will be existent upon initialization of in-memory catalog. What do you think?

Reply:
I agree we should not get pulled into the discussion around yaml file.

I still believe thought from the clean code perspective we should allow for passing a CatalogDatabase (or GenericInMemoryCatalogDatabse if we want to limit the scope) in the ctor. Such approach helps to write a testable code(http://misko.hevery.com/attachments/Guide-Writing%20Testable%20Code.pdf or  http://misko.hevery.com/code-reviewers-guide/) (which in turn is easier to use, as tests also verify that the interfaces are usable). I see no reason why we should not allow for that. 

But in the end I will leave it for your call.

Reply:
Since we will have PR coming for YAML file structure for catalogs, let's defer the discussion until we see it. Now as far as I am concerned, I don't see a use case for a user to create a default DB before creating a catalog. It's also common for a db product to create default db automatically. If user needs there specific databases, they can create their own (possibly via DDL). Our in-memory catalog is following the suite except we allow user specific a name other than "default" with the change here. This addition seems to me even optional. Nevertheless, if we find it necessary that taking a database created by user to instantiate the in-memory DB, we can add it at that time.